### PR TITLE
Polars: horizontal stack

### PR DESCRIPTION
**Pull request by Shoeboxam**

### CODE REVIEW

**Comment by Shoeboxam** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/opendp/opendp/1592?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1592** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1592?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1591** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1591?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1534** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1534?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1539** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1539?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1540** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1540?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1586** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1586?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1590** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1590?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1584** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1584?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1542** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1542?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1498** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1498?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1479** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1479?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1496](https://github.com/opendp/opendp/pull/1496) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1496?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#908** <a href="https://app.graphite.dev/github/pr/opendp/opendp/908?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#913](https://github.com/opendp/opendp/pull/913) <a href="https://app.graphite.dev/github/pr/opendp/opendp/913?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#890** <a href="https://app.graphite.dev/github/pr/opendp/opendp/890?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#856](https://github.com/opendp/opendp/pull/856) <a href="https://app.graphite.dev/github/pr/opendp/opendp/856?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#829** <a href="https://app.graphite.dev/github/pr/opendp/opendp/829?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#854](https://github.com/opendp/opendp/pull/854) <a href="https://app.graphite.dev/github/pr/opendp/opendp/854?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#888](https://github.com/opendp/opendp/pull/888) <a href="https://app.graphite.dev/github/pr/opendp/opendp/888?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1519** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1519?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1506](https://github.com/opendp/opendp/pull/1506) <a href="https://app.graphite.dev/github/pr/opendp/opendp/1506?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1566** <a href="https://app.graphite.dev/github/pr/opendp/opendp/1566?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Shoeboxam and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    // create a transformation for each expression
+    let expr_domain = ExprDomain::new(middle_domain.clone(), ExprContext::RowByRow);
+    let t_exprs = exprs
+        .into_iter()
+        .map(|expr| expr.make_stable(expr_domain.clone(), middle_metric.clone()))
+        .collect::<Fallible<Vec<_>>>()?;
+
+    // expand and update the set of series domains on the output domain
+    let mut series_domains = Vec::new();
+    let mut lookup = HashMap::new();
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs:54]:
```suggestion
    let mut name_to_index = HashMap::new();
```
Or a comment explaining the keys and values that are going into `lookup`. Explicit typing might also be used to make it clear what the keys and values are going to be, but just human readable would be fine.

File: rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        });
+
+    // only keep margins for series that have not changed
+    let new_series_names = new_series
+        .map(|series_domain| series_domain.field.name.to_string())
+        .collect();
+    let margins = (middle_domain.margins.iter())
+        .filter(|(k, _)| k.is_disjoint(&new_series_names))
+        .map(|(k, v)| (k.clone(), v.clone()))
+        .collect();
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs:81]: might be pulled into a tiny helper function, but the comment is probably sufficient.

File: rust/src/transformations/make_stable_lazyframe/h_stack/test.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    domains::{AtomDomain, LazyFrameDomain, Margin, OptionDomain, SeriesDomain},
+    metrics::SymmetricDistance,
+    transformations::make_stable_lazyframe,
+};
+
+use super::*;
+
+#[test]
+fn test_with_column() -> Fallible<()> {
+    let lf = df!("chunk_2_null" => [[Some(1i64); 500], [None; 500]].concat())?.lazy();
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_lazyframe/h_stack/test.rs:11]: Same comment from earlier PR: Does 500 exercise something that just 5 wouldn't?

File: rust/src/transformations/make_stable_lazyframe/h_stack/test.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        lf_domain.clone(),
+        SymmetricDistance,
+        lf.clone().fill_nan(lit(2.)),
+    )?;
+
+    let actual = t_with_column.invoke(&lf)?.collect()?;
+    let expect = df!("f64" => [[1f64; 500], [2f64; 500]].concat())?;
+
+    assert_eq!(actual, expect);
+    Ok(())
```
Comment (mccalluc) [inline rust/src/transformations/make_stable_lazyframe/h_stack/test.rs:60]: This is a reasonable thing to check, but the connection between `fill_nan` and `hstack` or `with_column` isn't clear to me.

**Review by mccalluc — COMMENTED** [review]
Comments, but it could be mostly my confusion.

File: rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    Expr: StableExpr<M, M>,
+{
+    let LogicalPlan::HStack {
+        input,
+        exprs,
+        schema,
+        options,
+    } = plan
+    else {
+        return fallible!(MakeTransformation, "Expected with_columns logical plan");
```
[the remark this answers is not in the document]
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_lazyframe/h_stack/mod.rs:39]: DataFrame has hstack, but LazyFrame doesn't. However, LazyFrame's with_columns makes a HStack node in the internal DSL.

File: rust/src/transformations/make_stable_lazyframe/h_stack/test.rs (hunk 4 above)
Comment (Shoeboxam) [reply rust/src/transformations/make_stable_lazyframe/h_stack/test.rs:60]: There's a comment above the test ^^

**Pull request merged**